File: LeetCode/79WordSearch/WordSearch.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        # DFS + Backtracking -> Time:O(n * m * 4^n), Space:O(m)
        visited = set()
        def backtrack(r, c, curr):
            if curr == len(word):
                return True
            if (
                r < 0
                or c < 0
                or r >= len(board)
                or c >= len(board[0])
                or word[curr] != board[r][c]
                or (r, c) in visited
            ):
                return False
            visited.add((r, c))
            res = (
                backtrack(r + 1, c, curr + 1)
                or backtrack(r - 1, c, curr + 1)
                or backtrack(r, c + 1, curr + 1)
                or backtrack(r, c - 1, curr + 1)
            )
            visited.remove((r, c))
            return res

        for r in range(len(board)):
            for c in range(len(board[0])):
                if backtrack(r, c, 0):
                    return True
        return False

File: LeetCode/79WordSearch/WordSearch.cpp
class Solution {
public:
    bool backtrack(vector<vector<char>>& board, string &word, set<vector<int>> &visited, int r, int c, int curr) {
        if (curr == word.size()) {
            return true;
        }
        vector<int> index;
        index.push_back(r);
        index.push_back(c);
        if (r < 0
            || c < 0
            || r >= board.size()
            || c >= board[0].size()
            || word[curr] != board[r][c]
            || visited.find(index) != visited.end()) {
                return false;
            }
        visited.insert(index);
        bool res = (backtrack(board, word, visited, r + 1, c, curr + 1)
                    || backtrack(board, word, visited, r - 1, c, curr + 1)
                    || backtrack(board, word, visited, r, c + 1, curr + 1)
                    || backtrack(board, word, visited, r, c - 1, curr + 1));
        visited.erase(index);
        return res;
    }
    bool exist(vector<vector<char>>& board, string word) {
        // DFS + Backtracking -> Time:O(n * m * 4^n), Space:O(m)
        set<vector<int>> visited;
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (backtrack(board, word, visited, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
};
